build/scripts/package.ts

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// @ts-nocheck <-- This is needed as this file is basically a JavaScript script
+//                 but with some TypeScript niceness baked in
+const fs = require('fs')
+const path = require('path')
+const glob = require('glob')
+const execSync = require('child_process').execSync
+const pkg_dir = 'dist/package'
+
+async function copyGlob(pattern) {
+  glob(pattern, (error, files) => {
+    for (var i = 0; i < files.length; i++) {
+      // Skip dirs as they are most likely listed in the .vscodeignore as well
+      if (fs.statSync(files[i]).isDirectory()) continue
+
+      let src = files[i]
+      let dst = path.join(pkg_dir, src.replace('build/package', ''))
+      let dstDir = path.dirname(dst)
+
+      if (!fs.existsSync(dstDir)) {
+        fs.mkdirSync(dstDir, { recursive: true })
+      }
+
+      fs.copyFileSync(src, dst)
+    }
+  })
+}
+
+// Setup package directory
+function setup() {
+  if (fs.existsSync(pkg_dir)) {
+    fs.rmdirSync(pkg_dir, { recursive: true })
+  }
+
+  fs.mkdirSync(pkg_dir)
+
+  let lines = fs
+    .readFileSync('build/package/.vscodeignore')
+    .toString()
+    .split('\n')
+
+  // Copy all files listed in the .vscodeignore
+  for (var i = 0; i < lines.length; i++) {
+    var line = lines[i]
+
+    if (!line.startsWith('!')) continue
+
+    let pattern = line.substring(1).trim()
+
+    copyGlob(pattern)
+  }
+
+  // Copy required package files into package directory
+  copyGlob('build/package/{.,}*') // include hidden files
+
+  fs.copyFileSync('yarn.lock', `${pkg_dir}/yarn.lock`)
+}
+
+// Create VSIX package
+function create() {
+  execSync('yarn install', { cwd: pkg_dir })
+  execSync('yarn vsce package --out ../../', { cwd: pkg_dir })
+}
+
+module.exports = {
+  setup: setup,
+  create: create,
```

Comment:
Just out of curiosity, previously you had the exec stuff done after files were copied and now you have it split into separate functions. This is fine with me, but I'm trying to get a better grasp of how yarn works--is there a special reason this is needed, or just thought it would be good to split things up?

Reply:
So it is split because when I try to run those `execSync` commands in the same method as the copies, those cmds try to run before everything is copied causing package issues. So splitting out was done to make sure the package directory is completely setup before actually trying to run the package commands

Reply:
Yeah since this TypeScript is nodejs script not sure if the async/await works completely